### PR TITLE
fix: fix migrate different datasources on same project

### LINKED ISSUE #5479: Migrating MySQL and Mongo on the same project causes a crash
I followed [this tutorial](https://loopback.io/doc/en/lb4/todo-list-tutorial-sqldb.html) and everything worked fine. I managed to work with relations on MySQL db.

In my use case, i had to add another datasource based on Mongo and i had a bug while migrating

> Unable to migrate models not associated with this data source

i had to translate it from french, the original text is 

> Impossible de migrer les modèles non associés à cette source de données

i digged a little in `RepositoryMixin` class, and found out that the problem came from `migrateSchema` method

i overrided the method (copy past x) ) and changed 

```
for (const b of dsBindings) {
  const ds = await this.get(b.key);
  if (operation in ds && typeof ds[operation] === 'function') {
    debug('Migrating dataSource %s', b.key);
    await ds[operation](options.models);
  }
  else {
    debug('Skipping migration of dataSource %s', b.key);
  }
}
```

to 

```
for (const b of dsBindings) {
      const ds = await this.get(b.key);
      // console.log(ds);
      // @ts-ignore
      if (operation in ds && typeof ds[operation] === 'function') {
        console.log('Migrating dataSource %s', b.key);
        // @ts-ignore
        if (!options?.models || options?.models.length === 0) {
          // @ts-ignore
          await ds[operation](options.models);
        } else {
          const models = [];
          for (const model of options?.models) {
            // @ts-ignore
            if (Object.keys(ds.definitions).includes(model)) {
              models.push(model);
            }
          }
          // @ts-ignore
          await ds[operation](models);
        }
      } else {
        // debug('Skipping migration of dataSource %s', b.key);
      }
    }
```

i hope this bug will be fixed soon

### PR DESCRIPTION
**Pull request by khlilturki97**
this commit fixes bug reported in issue 5479

fix #5479

## Checklist

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by raymondfeng** [comment]
@khlilturki97 Thank you for the patch. Please fix the compilation error as reported by CI:

```
packages/repository/src/mixins/repository.mixin.ts:303:31 - error TS7053: Element implicitly has an 'any' type because expression of type '"definitions"' can't be used to index type 'DataSource'.
  Property 'definitions' does not exist on type 'DataSource'.
303               if (Object.keys(ds['definitions']).includes(model)) {
                                  ~~~~~~~~~~~~~~~~~
```

**Comment by khlilturki97** [comment]
There's an attribute that needs to be added in DataSource model in "loopback-datasource-juggler" 

I dont know how to fix it

File: packages/repository/src/mixins/repository.mixin.ts [hunk 1]
```diff
@@ -289,7 +290,22 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
...
+            // if options.models is an empty array, proceed normally
+            await ds[operation](options.models);
+          } else {
+            // if options.models is not empty, check if model is
+            // connected to the datasource
+            const models = [];
+            for (const model of options?.models) {
+              // todo : add "definitions" attribute in Datasource definition
+              // in "loopback-datasource-juggler"
+              if (Object.keys(ds['definitions']).includes(model)) {
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:303]: > There's an attribute that needs to be added in DataSource model in "loopback-datasource-juggler"

Here is the place where to add type declarations for `definitions` property:

https://github.com/strongloop/loopback-datasource-juggler/blob/8d1690e9b2c76e3d0a8336b47f133e0605cbcc9a/types/datasource.d.ts#L77-L85

File: packages/repository/src/mixins/repository.mixin.ts [hunk 2]
```diff
@@ -289,7 +290,22 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
...
+                models.push(model);
+              }
+            }
+            await ds[operation](models);
+          }
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:308]: I find the current proposal a bit more complex than I like. How about structuring the code as follows?

```ts
// here, inside migrateSchema()
const models = filterDataSourceModels(options.models, ds);
await ds[operation](models);

// a helper implemented outside of the mixin class
function filterDataSourceModels(models, dataSource) {
  if (!models || models.length === 0) return models;
  const definitions = Object.keys(ds.definitions);
  return models.filter(m => definitions.includes(m));
}
``

File: examples/access-control-migration/src/repositories/user.repository.ts [hunk 3]
```diff
@@ -59,7 +59,7 @@ export class UserRepository extends DefaultCrudRepository<
     userId: typeof User.prototype.id,
   ): Promise<UserCredentials | undefined> {
     try {
-      return await this.userCredentials(userId).get();
+      return this.userCredentials(userId).get();
```
[the remark this answers is not in the document]
Comment (khlilturki97) [reply examples/access-control-migration/src/repositories/user.repository.ts:62]: i think the linter changed this, i will fix it

File: packages/repository/src/mixins/repository.mixin.ts [hunk 4]
```diff
@@ -270,6 +268,9 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
      * preserving data or rebuild everything from scratch.
      */
     async migrateSchema(options: SchemaMigrationOptions = {}): Promise<void> {
+      if (options.existingSchema === 'drop' && options?.models) {
```
[the remark this answers is not in the document]
Comment (khlilturki97) [reply packages/repository/src/mixins/repository.mixin.ts:271]: this shows a compilation error, it says that options might not be defined, the "?" solved the problem

File: packages/repository/src/mixins/repository.mixin.ts (hunk 1 above)
Comment (khlilturki97) [reply packages/repository/src/mixins/repository.mixin.ts:303]: i have to create a PR to fix it?

File: packages/repository/src/mixins/repository.mixin.ts (hunk 2 above)
Comment (khlilturki97) [reply packages/repository/src/mixins/repository.mixin.ts:308]: it's easier to read, true

i'll test it

**Comment by khlilturki97** [comment]
@bajtos @raymondfeng i have commited new changes

also i have created a [PR](https://github.com/strongloop/loopback-datasource-juggler/pull/1844) in [loopback datasource juggler](https://github.com/strongloop/loopback-datasource-juggler)

File: packages/repository/src/mixins/repository.mixin.ts (hunk 1 above)
Comment (bajtos) [reply packages/repository/src/mixins/repository.mixin.ts]: Yes please!

Comment (bajtos) [reply packages/repository/src/mixins/repository.mixin.ts]: I see you opened https://github.com/strongloop/loopback-datasource-juggler/pull/1844 👏🏻

File: packages/repository/src/mixins/repository.mixin.ts [hunk 5]
```diff
@@ -25,10 +25,6 @@ import {juggler, Repository} from '../repositories';
 
 const debug = debugFactory('loopback:repository:mixin');
 
-// FIXME(rfeng): Workaround for https://github.com/microsoft/rushstack/pull/1867
-/* eslint-disable @typescript-eslint/no-unused-vars */
-import * as loopbackContext from '@loopback/core';
-import * as loopbackCore from '@loopback/core';
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:31]: Please revert the changes above. IDE tooling is sorting imports in a way that breaks our API doc generator, we have to manually preserve the explicit order.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 6]
```diff
@@ -139,14 +135,12 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
       const options = toOptions(nameOrOptions);
       // We have an instance of
       if (dataSource instanceof juggler.DataSource) {
-        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
-        const name = options.name || dataSource.name;
+        const name = options.name ?? dataSource.name;
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:138]: I believe we are intentionally using the operator `||` here, please revert this change.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 7]
```diff
@@ -139,14 +135,12 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
...
         const namespace = options.namespace ?? RepositoryBindings.DATASOURCES;
         const key = `${namespace}.${name}`;
         return this.bind(key).to(dataSource).tag(RepositoryTags.DATASOURCE);
       } else if (typeof dataSource === 'function') {
-        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
-        options.name = options.name || dataSource.dataSourceName;
+        options.name = options.name ?? dataSource.dataSourceName;
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:143]: Ditto.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 8]
```diff
@@ -336,15 +366,21 @@ export interface ApplicationWithRepositories extends Application {
...
     name?: string,
   ): Binding<D>;
+
   model<M extends Class<unknown>>(modelClass: M): Binding<M>;
+
   component(component: Class<unknown>, name?: string): Binding;
+
   mountComponentRepositories(component: Class<unknown>): void;
+
   migrateSchema(options?: SchemaMigrationOptions): Promise<void>;
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:384]: Please revert the addition of new lines above, it seems to me as unrelated to fixing datasource migration.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 9]
```diff
@@ -467,7 +503,8 @@ export class RepositoryMixinDoc {
    *
    * @param component - The component to mount repositories of
    */
-  mountComponentRepository(component: Class<{}>) {}
+  mountComponentRepository(component: Class<{}>) {
+  }
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:507]: Ditto, please revert this unrelated change.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 10]
```diff
@@ -482,7 +519,8 @@ export class RepositoryMixinDoc {
    * @param options - Migration options, e.g. whether to update tables
    * preserving data or rebuild everything from scratch.
    */
-  async migrateSchema(options?: SchemaMigrationOptions): Promise<void> {}
+  async migrateSchema(options?: SchemaMigrationOptions): Promise<void> {
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:522]: Ditto, please revert this unrelated change.

File: packages/repository/src/mixins/repository.mixin.ts [hunk 11]
```diff
@@ -289,12 +286,45 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
...
         } else {
           debug('Skipping migration of dataSource %s', b.key);
         }
       }
     }
+
+    /**
+     * Check if a model is part of datasource
+     *
+     * Used by migrate schema
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:317]: I find this comment as a bit misleading. When you say "check if ...", I would expect the function to return `true` or `false` based on the check outcome.

I also feel it's not necessary to explain where is this function used - this fact can change in the future, plus it's easy to find using "find all references" or a string-based search.

Proposed text:

```suggestion
     * Filter the list of models to keep only the models attached to the given datasource.
```

File: packages/repository/src/mixins/repository.mixin.ts [hunk 12]
```diff
@@ -289,12 +286,45 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
...
+          //   const models = [];
+          //   for (const model of options.models) {
+          //     // todo : add "definitions" attribute in Datasource definition
+          //     // in "loopback-datasource-juggler"
+          //     if (Object.keys(ds['definitions']).includes(model)) {
+          //       models.push(model);
+          //     }
+          //   }
+          //   await ds[operation](models);
+          // }
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:307]: Please remove the commented-out code.

File: packages/repository/src/mixins/repository.mixin.ts (hunk 5 above)
Comment (khlilturki97) [reply packages/repository/src/mixins/repository.mixin.ts:31]: i didn't get your point, i don't see those imports

File: packages/repository/src/mixins/repository.mixin.ts (hunk 11 above)
Comment (khlilturki97) [reply packages/repository/src/mixins/repository.mixin.ts:317]: thanks for the description i was stuck ^^

File: packages/repository/src/mixins/repository.mixin.ts (hunk 5 above)
Comment (khlilturki97) [reply packages/repository/src/mixins/repository.mixin.ts:31]: i solved it, i didn't see that it was removed

File: packages/repository/src/mixins/repository.mixin.ts [hunk 13]
```diff
@@ -3,17 +3,17 @@
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
 
+// FIXME(rfeng): Workaround for https://github.com/microsoft/rushstack/pull/1867
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:7]: Please remove these two lines, they should stay at L28 & L29 only (see below in this file).

File: packages/repository/src/mixins/repository.mixin.ts [hunk 14]
```diff
@@ -139,13 +139,11 @@ export function RepositoryMixin<T extends MixinTarget<Application>>(
       const options = toOptions(nameOrOptions);
       // We have an instance of
       if (dataSource instanceof juggler.DataSource) {
-        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
         const name = options.name || dataSource.name;
         const namespace = options.namespace ?? RepositoryBindings.DATASOURCES;
         const key = `${namespace}.${name}`;
         return this.bind(key).to(dataSource).tag(RepositoryTags.DATASOURCE);
       } else if (typeof dataSource === 'function') {
-        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:148]: I think you will need to keep those `eslint-disable-next-line` comments to make `npm run lint` pass. Unfortunately, the linter is chocking at TypeScript errors caused by the fact that we don't have https://github.com/strongloop/loopback-datasource-juggler/pull/1844 yet. Feel free to wait with changes until that other PR is landed & published and loopback-next picks up the new version.

**Comment by bajtos** [comment]
#6008 updated package & package-lock to use the new juggler, can you please rebase your pull request on top of the latest master and address the remaining review comments?

**Comment by khlilturki97** [comment]
@bajtos done rebasing

**Comment by dhmlau** [comment]
@khlilturki97, CI failed with the error below. Could you please take a look? Thanks.
```
packages/repository/src/mixins/repository.mixin.ts(52,17): error TS4094: Property 'filterDataSourceModels' of exported class expression may not be private or protected.
```

**Comment by khlilturki97** [comment]
@dhmlau does changing 'filterDataSourceModels' to public resolve the problem.?

**Comment by dhmlau** [comment]
We just switch the contribution method from CLA to DCO, making your contribution easier in the future. Please sign the commits with DCO by amending your commit messages with `-s` flag and push the changes again. If you're using the command line, you can:

```
git commit --amend -s
git push --force-with-lease
```

Please refer to [this docs page](https://loopback.io/doc/en/contrib/code-contrib.html#developer-certificate-of-origin-dco) for details. Thanks!

**Comment by dhmlau** [comment]
@khlilturki97, there are multiple code linting errors. Please run `npm run lint` to find out more about the errors.

**Comment by bajtos** [comment]
@khlilturki97 

> does changing 'filterDataSourceModels' to public resolve the problem.?

It may solve it. You can try it yourself - run `npm test` from `loopback-next` root to verify.

**Comment by bajtos** [comment]
@khlilturki97 Ping, have you had a chance to look at the test/lint failures? Is there anything we can help you with?

**Comment by raymondfeng** [comment]
@khlilturki97 I went ahead to fix the TSC compilation error and squashed all commits into one. Please acknowledge the change.

**Comment by khlilturki97** [comment]
@raymondfeng can you show me how ?

File: packages/repository/src/mixins/repository.mixin.ts [hunk 15]
```diff
@@ -25,10 +23,6 @@ import {juggler, Repository} from '../repositories';
 
 const debug = debugFactory('loopback:repository:mixin');
 
-// FIXME(rfeng): Workaround for https://github.com/microsoft/rushstack/pull/1867
-/* eslint-disable @typescript-eslint/no-unused-vars */
-import * as loopbackContext from '@loopback/core';
-import * as loopbackCore from '@loopback/core';
```
Comment (bajtos) [inline packages/repository/src/mixins/repository.mixin.ts:31]: @raymondfeng are you sure this workaround is no longer needed?
